Step04/app/src/main/java/com/kwoolytech/step04/MainActivity.java

```diff
@@ -0,0 +1,168 @@
+        httpClientCallback = new KwoolyHttpCallback() {
+            @Override
+            public void onWeatherJsonDataReceived(String result) {
+                try {
+                    dataModel.setWeatherJsonData(result);
+                    queryWeatherBitmapData();
+                } catch (Exception e) {
+                    Log.e(getClass().getName(), "Exception: ");
+                    e.printStackTrace();
+                }
+            }
+
+            @Override
+            public void onWeatherBitmapDataReceived(Bitmap result) {
+                try {
+                    dataModel.setWeatherBitmapData(result);
+                    presentData();
+                } catch (Exception e) {
+                    Log.e(getClass().getName(), "Exception: ");
+                    e.printStackTrace();
+                }
+            }
+        };
+    }
+
+    private void queryWeatherJsonData() {
+        try {
+            httpClient.httpGetJsonData(
+                    CommonTool.WeatherQueryUrl + "lat=37.49&lon=127.01&units=metric" + CommonTool.ApiKey,
+                    httpClientCallback);
+        } catch (Exception e) {
+            Log.e(getClass().getName(), "Exception: ");
+            e.printStackTrace();
+        }
+        return;
+    }
+
+    private void queryWeatherBitmapData() {
+        try {
+            httpClient.httpGetBitmapData(
+                    CommonTool.WeatherIconUrl + dataModel.getIconCode() + ".png",
+                    httpClientCallback);
+        } catch (Exception e) {
+            Log.e(getClass().getName(), "Exception: ");
+            e.printStackTrace();
+        }
+        return;
+    }
+
+    private void presentData() {
+        listViewWeatherAdapter.notifyDataSetChanged();
+        return;
+    }
+
+    @Override
+    public void onBackPressed() {
+        DrawerLayout drawer = (DrawerLayout)findViewById(R.id.drawer_layout);
+        if (drawer.isDrawerOpen(GravityCompat.START)) {
+            drawer.closeDrawer(GravityCompat.START);
+        } else {
+            super.onBackPressed();
+        }
+    }
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.main, menu);
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        int id = item.getItemId();
+
+        if (id == R.id.action_settings) {
+            return true;
+        }
+
+        return super.onOptionsItemSelected(item);
+    }
+
+    @SuppressWarnings("StatementWithEmptyBody")
+    @Override
+    public boolean onNavigationItemSelected(MenuItem item) {
+        int id = item.getItemId();
+
+        if (id == R.id.nav_camera) {
+
+        } else if (id == R.id.nav_share) {
+
+        } else if (id == R.id.nav_send) {
+
+        }
+
+        DrawerLayout drawer = (DrawerLayout)findViewById(R.id.drawer_layout);
+        drawer.closeDrawer(GravityCompat.START);
+
```

Comment:
I thought that closing drawer after selecting any of items is handled automatically. But it should be closed manually. Yes it makes sense to close drawer and depends on use-cases per item.